server/fleet/app.go

```diff
@@ -239,6 +273,44 @@ func (c *AppConfig) ApplyDefaults() {
 	c.WebhookSettings.Interval.Duration = 24 * time.Hour
 }
 
+// EnableStrictDecoding enables strict decoding of the AppConfig struct.
+func (c *AppConfig) EnableStrictDecoding() { c.strictDecoding = true }
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (c *AppConfig) UnmarshalJSON(b []byte) error {
+	// Define a new type, this is to prevent infinite recursion when
+	// unmarshalling the AppConfig struct.
+	type cfgStructUnmarshal AppConfig
+	compatConfig := struct {
+		*legacyConfig
+		*cfgStructUnmarshal
+	}{
+		&legacyConfig{},
+		(*cfgStructUnmarshal)(c),
+	}
+
+	decoder := json.NewDecoder(bytes.NewReader(b))
+	if c.strictDecoding {
+		decoder.DisallowUnknownFields()
+	}
+	if err := decoder.Decode(&compatConfig); err != nil {
+		return err
+	}
+	if _, err := decoder.Token(); err != io.EOF {
+		return errors.New("unexpected extra tokens found in config")
+	}
+
+	// TODO(roperzh): define and assign legacy settings to new fields. This has
+	// the drawback of legacy fields taking precedence over new fields if both
```

Comment:
that's an excellent point, will adjust the related PR that actually uses this method.

Reply:
@michalnicp we would have to make the new fields pointers for that to work, right? I wonder if it's worth the hassle

Reply:
If we make sure that when the appconfig is saved again, legacy fields are never saved (because anyway their values got transferred to the new fields), then having both new and legacy should never happen, and so this approach would be fine as-is? I believe this is what would happen with this implementation, correct me if I'm wrong (i.e. `legacyConfig` is not part of `AppConfig` and so would never get marshalled+saved).

Reply:
that is correct, and I think I should make the comment in the code cleaner, but what I mean is that if an user specifies _both_ a legacy config and a new config when _applying_ a config, the legacy wins:

```
~/fleet $ cat /tmp/cfg.yml
apiVersion: v1
kind: config
spec:
   features:
     additional_queries:
       - "this is new"
   host_settings:
     additional_queries:
       - "this is old"

~/fleet $ ./build/fleetctl apply -f /tmp/cfg.yml
[+] applied fleet config

~/fleet $ ./build/fleetctl get config --json | jq '.spec.features'
{
  "enable_host_users": false,
  "enable_software_inventory": false,
  "additional_queries": [
    "this is old"
  ]
}
```

Reply:
Oh I see. There's this ticket we have that is coming up soon-ish: https://github.com/fleetdm/fleet/issues/5222

It would make a lot of sense to reject configs with legacy fields when we implement this, and that would alleviate the concern here. Of course there's still the case where the user force-applies a config, so it might still make sense to avoid the legacy-wins-over-new.

Reply:
thanks for the context on that, given that:

- if you only apply a legacy field, we're good
- if you only apply a new field, we're good
- if you apply both fields at the same time, legacy wins

the validation could be then that you can't specify both a legacy field + a new field at the same time. If despite all of that, you specify `--force`, then I feel OK with this behavior as it's such an edge case.

Reply:
Any reason why we should allow a legacy field at all, even if the new is not set? Keeping in mind that it's the fleet server instance that would do the validation (i.e. not fleetctl), so if the fleet instance knows that a field is legacy, it's because it knows about the new field, so there's no good reason to accept the legacy one

> If despite all of that, you specify --force, then I feel OK with this behavior as it's such an edge case.

Yeah I'd agree with that too.

Reply:
No reason on my end! this epic specified to support both `host_settings` (legacy) and `features` (new) keys, but I guess that if there's an error message when you specify we're good!

Only downside I see is that users will have to update all their configs as soon as the validation feature is implemented. Maybe we could allow both for a while? The whole deprecation/removal process definitely needs some discussion outside of both issues (this and https://github.com/fleetdm/fleet/issues/5222) I think

Reply:
Definitely, thanks for the feedback and the example scenario, I'll make sure to discuss this with Noah when I start work on the validation part.